### PR TITLE
Implement --prefer-oldest

### LINKED ISSUE #2876: Parameter to make solver pick the minimal set of version depencencies
This would be useful to detect wrong minimum dependency version bounds.

@kosmikus, would this be too difficult?

### PR DESCRIPTION
**Pull request by Bodigrim**
As suggested in https://github.com/haskell/cabal/issues/2876#issuecomment-1172072384. 

If this looks sensible, I can add tests / documentation / changelog.

Closes #2876.

### CODE REVIEW

**Comment by Bodigrim** [comment]
Any feedback please?

**Comment by jneira** [comment]
It looks fine to me at a first glance, it makes the interface more balanced and comments alone are great. 
Otoh i have the impression participants in the issue discussion was in favour of doing it (orat least was not against)
but i would like to know what other maintainers think about: @grayjay @fgaz @Mikolaj

**Review by Mikolaj — CHANGES_REQUESTED** [review]
This looks clean and scratches a real life itch. If I'm not mistaken, the price for the new functionality is neither a significant complication nor a slowdown of the existing code. However, if I am mistaken (and I don't know this code well enough), let's try to remedy this.

A test and a changelog would be much appreciated.

**Comment by Bodigrim** [comment]
@Mikolaj done.

**Review by Mikolaj — APPROVED** [review]
Looks good, thank you.

**Comment by jneira** [comment]
tbh i was expecting some sort of integration test in the cabal-testsuite, which hopefully could serve to see a functional example of the feature (a freeze file for example?). But i am not sure if it would be feasible, and there are unit tests so i would be fine with them.

It would be great to have some sort of documentation about but i cant even find the prior documentation 🤷

**Comment by Bodigrim** [comment]
@jneira thanks, I've added the flag to the user's manual. Is it all set now?

**Review by jneira — APPROVED** [review]
yeah, it looks great, code comments and docs included, thanks!
I would wait some time to see @grayjay has the chance to take a look before merge if you don't mind

**Comment by Bodigrim** [comment]
Sure, but I'd really love to get this merged in time to be backported into 3.8 release.

**Comment by jneira** [comment]
yeah, not sure what is the plan for the release, @Mikolaj will you cut another prerelease before the definitive one?

**Comment by grayjay** [comment]
Sorry I haven't had a chance to review this yet. I'd also like to try out running it on real packages. I think I'll have time to review it by Wednesday.

**Comment by Bodigrim** [comment]
@grayjay I tried `--prefer-oldest` on `haskell-language-server` and filed a bunch of PRs with lower bounds.

**Comment by Mikolaj** [comment]
Please take your time. That's not the last release of cabal in history, I hope.

Re pre-releases, we'll decide on Thursday, but I will argue for another RC, due to the process interrupt snags and due to building with GHC 9.2 (which we should have switched to earlier). Then, with some luck, we'd release RC2 without any changes (except version numbers) as 3.8-final.

Regardless, we are certainly not eager to backport a non-fix, non-doc PR at this late stage. @Bodigrim: how much trouble would it be for you to use master branch (we can easily make an unofficial release via gitlab-ci) until cabal 3.10?

**Comment by Bodigrim** [comment]
@Mikolaj Not the end of the world, but a considerable obstacle. To gain the most value of this feature, it needs to be run as a CI job, potentially via `haskell-ci`. Obviously, the later would not accept a PR depending on an unreleaded version of Cabal. Any wide community adoption would not happen as well.

From my perspective this is an isolated feature, which is entirely opt-in and does not interfere with the rest of Cabal. Changing the bootstrap compiler or the `process` version are way more dangerous and potentially disrupting things.

That said, I'm in no position to interfere with your release cycles, no hard feelings.

**Comment by Mikolaj** [comment]
Some code changes in the PR are pretty intrusive but most of these are in the test code, so that's actually in favour of backporting. If we had many features clamouring to get backported, that would be a hard no, but you are the only one that tries to bribe us into a late backport, so let's see what the team says on Thursday.

**Comment by Bodigrim** [comment]
@grayjay just a gentle reminder to review.

**Review by grayjay — COMMENTED** [review]
This feature looks good overall. Thanks for adding it!

I'm not sure I understand the interaction between --prefer-oldest and the `PreferLatestForSelected` preference default. It would help to add a comment somewhere explaining how they relate.

With the addition of --prefer-oldest, I think it would be clearer if `PreferLatest` were named `PreferNewest`. I think that's fine, though, because the `PreferLatest` wording has never been exposed. I'm not suggesting renaming anything now.

**Comment by Bodigrim** [comment]
@grayjay fixed. Any more suggestions?

**Review by grayjay — APPROVED** [review]
Thanks!

**Comment by Bodigrim** [comment]
@Mikolaj good to go?

**Comment by Mikolaj** [comment]
Let's go! merge_me or squash+merge_me, your pick.

**Comment by Bodigrim** [comment]
> @Bodigrim: after today's meeting, it seems to me the risk appetite in the cabal devs community justifies merging this into a 3.8 point release (if there is any; there probably will be, unless 3.10 is very prompt), but we'd rather not squeeze it into the first 3.8 build.

Which labels are justified in this case? Just `attention: needs-backport 3.8`, but no actual backport yet?

**Comment by Mikolaj** [comment]
I've created a label especially for this PR. :)

**Comment by phadej** [comment]
FYI: There should be a disclaimer in the docs. Depending on which order solver picks packages it may pick smaller version of pkg `A`, and then can only pick bigger version of `B`. If it traversed packages in opposite order, that plan would be different.

So `--prefer-oldest` cannot find *the best* install-plan in general (as cannot `--prefer-latest` either).
I suspect that with oldest version the landscape is much more commonly not convex than with default, as by default people try to support latest versions of their dependencies.

The best (but slow) method to find the lowest supported dependency versions is to go them one by one and figure out which is the smallest version. (I do it by binary search solving multiple times, having `--prefer-oldest-for-pkg=foo` would be better resource wise).

I kindly ask to write more prose in the documentation, as general audience may not understand the limitations of `--prefer-oldest` flag.

**Comment by Mikolaj** [comment]
@Bodigrim: could you comment on the remark above?

**Comment by Bodigrim** [comment]
@Mikolaj how do you feel about backporting this to 3.8 branch now, after 3.8.1.0 release?

**Comment by Mikolaj** [comment]
Let's do it.

**Comment by Mikolaj** [comment]
@mergify backport 3.8

**Comment by Mikolaj** [comment]
@Bodigrim: merged to branch 3.8. Thank you.

File: cabal-install-solver/src/Distribution/Solver/Types/InstalledPreference.hs
Comment (phadej) [inline cabal-install-solver/src/Distribution/Solver/Types/InstalledPreference.hs:10]: Now as `cabal-install-solver` is a library, changes like this are **BREAKING CHNAGES**, so you don't simply backport them anymore to `cabal-install` releases.

I has warned about this problem, please deal with it appropriately. This change cannot be backported without a major version bump.

Comment (Mikolaj) [reply cabal-install-solver/src/Distribution/Solver/Types/InstalledPreference.hs:10]: I seem to remember `cabal-install-solver` designated as an "internal library", at least for the time being while API is shaking up. However, I don't see it described like that on Hackage, which perhaps should be fixed? I think either one or the other should be fixed.

Comment (fgaz) [reply cabal-install-solver/src/Distribution/Solver/Types/InstalledPreference.hs:10]: iirc in the end it was decided to honor pvp, and that's why there isn't any notice. i opened #8386 to revert the backport

Comment (Mikolaj) [reply cabal-install-solver/src/Distribution/Solver/Types/InstalledPreference.hs:10]: @Bodigrim: if 3.8.2.0 gets released and this PR is not included, please accept my apologies. I missed the API change and negotiated the inclusion believing the choice is 3.8.1.0 vs 3.8.2.0, not 3.8.1.0 vs 3.10.1.0. Perhaps we should just skip 3.8.2.0 regardless of anything and release 3.10.1.0 next, even if this PR is the only major change in it (but there is a lot of other changes accumulated on master already; I don't remember how breaking they are and it's to be seen if we'd base 3.10 off master or off current 3.8 with your API breaking change).

**Pull request merged**
